packages/host/tests/helpers/adapter.ts

```diff
@@ -0,0 +1,299 @@
+    path: string,
+  ): AsyncGenerator<{ name: string; path: string; kind: Kind }, void> {
+    let dir =
+      path === '' ? this.#files : this.#traverse(path.split('/'), 'directory');
+    for (let [name, content] of Object.entries((dir as Dir).contents)) {
+      yield {
+        name,
+        path: path === '' ? name : `${path}/${name}`,
+        kind: content.kind,
+      };
+    }
+  }
+
+  async exists(path: string): Promise<boolean> {
+    let maybeFilename = path.split('/').pop()!;
+    try {
+      // a quirk of our test file system's traverse is that it creates
+      // directories as it goes--so do our best to determine if we are checking for
+      // a file that exists (because of this behavior directories always exist)
+      await this.#traverse(
+        path.split('/'),
+        maybeFilename.includes('.') ? 'file' : 'directory',
+      );
+      return true;
+    } catch (err: any) {
+      if (err.name === 'NotFoundError') {
+        return false;
+      }
+      if (err.name === 'TypeMismatchError') {
+        try {
+          await this.#traverse(path.split('/'), 'file');
+          return true;
+        } catch (err: any) {
+          if (err.name === 'NotFoundError') {
+            return false;
+          }
+          throw err;
+        }
+      }
+      throw err;
+    }
+  }
+
+  async openFile(
+    path: LocalPath,
+    loader: Loader,
+  ): Promise<FileRef | undefined> {
+    let content;
+    try {
+      content = this.#traverse(path.split('/'), 'file');
+    } catch (err: any) {
+      if (['TypeMismatchError', 'NotFoundError'].includes(err.name)) {
+        return undefined;
+      }
+      throw err;
+    }
+    if (content.kind === 'directory') {
+      return undefined;
+    }
+
+    let value = content.content;
+
+    let fileRefContent = '';
+
+    if (path.endsWith('.json')) {
+      let cardApi = await loader.import<CardAPI>(`${baseRealm.url}card-api`);
+      if (cardApi.isCard(value)) {
+        value.id = `${this.#paths.url}${path.replace(/\.json$/, '')}`;
+        cardApi.setCardAsSavedForTest(value);
+        let doc = cardApi.serializeCard(value);
+        fileRefContent = JSON.stringify(doc);
+      } else {
+        fileRefContent =
+          typeof value === 'string' ? value : JSON.stringify(value);
+      }
+    } else if (path.endsWith('.gts')) {
+      if (typeof value === 'string') {
+        fileRefContent = value;
+      } else {
+        let moduleURLString = `${this.#paths.url}${path.replace(/\.gts$/, '')}`;
+
+        loader.shimModule(moduleURLString, value as object);
+
+        fileRefContent = shimmedModuleIndicator;
+      }
+    } else {
+      fileRefContent = value as string;
+    }
+
+    let fileRef: FileRef = {
+      path,
+      content: fileRefContent,
+      lastModified: this.#lastModified.get(this.#paths.fileURL(path).href)!,
+    };
+
+    if (fileRefContent === shimmedModuleIndicator) {
+      fileRef[Symbol.for('shimmed-module')] = value as object;
```

Comment:
We're passing the module's value as a symbol so that independent loaders can all get shimmed modules when requested.